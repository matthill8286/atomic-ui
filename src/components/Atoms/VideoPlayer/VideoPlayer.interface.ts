import React, { MutableRefObject, ReactElement, Ref } from 'react'

interface SourceProps {
  src: string
  type: string
}

export interface VideoPlayerViewer {
  onStart: (callback: () => void) => void
  loadUrl: (assetId: string) => void
  play: () => void
  stop: () => void
}

export interface VideoPlayerProps {
  title?: string
  unsupportedMedia?: boolean
  context?: boolean
  allowfullscreen?: boolean
  maxHeight?: string | number
  rounded?: boolean
  url: string
  loop?: boolean
  videoId?: string
  controls?: boolean
  loading?: boolean
  volume?: number
  muted?: boolean
  videoRef?: MutableRefObject<Ref<HTMLDivElement>>
  playbackRate?: number
  width?: string | undefined
  height?: string | undefined
  progressInterval?: number
  playsinline?: boolean
  playing?: boolean
  playIcon?: ReactElement
  previewTabIndex?: number
  pip?: boolean
  stopOnUnmount?: boolean
  light?: boolean | string
  fallback?: ReactElement
  onStart?: (callback: () => void) => void
  onPlay?: (event: React.MouseEvent<HTMLDivElement>) => void
  onPause?: (event: React.MouseEvent<HTMLDivElement>) => void
  onEnded?: (event: React.MouseEvent<HTMLDivElement>) => void
  onDuration?: (event: React.MouseEvent<HTMLDivElement>) => void
  onClickPreview?: (event: React.MouseEvent<HTMLDivElement>) => void
  onError?: (error: any, data?: any, hlsInstance?: any, hlsGlobal?: any) => void
  onProgress?: (state: {
    played: number
    playedSeconds: number
    loaded: number
    loadedSeconds: number
  }) => void
}
