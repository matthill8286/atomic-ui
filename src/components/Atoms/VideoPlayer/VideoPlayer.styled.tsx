import { css, styled } from '@/styles'
import ReactPlayer from 'react-player'

/**
 Aspect Ratio	Padding-Bottom
 1:1	                  100%
 16:9	                  56.25%
 4:3	                  75%
 3:2	                  66.66%
 8:5	                  62.5%
 **/

export const StyledVideoWrapper = styled.div<{ aspect: string }>(
  ({ aspect }) =>
    css`
      position: relative;
      height: 100%;
      width: 100%;

      :before {
        content: '';
        display: block;
        padding-bottom: calc(100% / calc(${aspect}));
      }

      > img {
        height: auto;
      }
    `
)

export const StyledVideoPlayer = styled(ReactPlayer)`
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  border: none;

  iframe {
    ${({ theme, rounded }) =>
      rounded ? `border-radius: ${theme.dimension.borderRadius2} !important;` : undefined};
  }
`
