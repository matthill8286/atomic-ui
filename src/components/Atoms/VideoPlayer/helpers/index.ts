export const MATCH_YOUTUBE = /(?:youtu\.be\/|youtube(?:-nocookie)?\.com\/(?:embed\/|v\/|watch\/|watch\?v=|watch\?.+&v=))((\w|-){11})|youtube\.com\/playlist\?list=|youtube\.com\/user\//
export const MATCH_SOUNDCLOUD = /(?:soundcloud\.com|snd\.sc)\/[^.]+$/
export const MATCH_VIMEO = /vimeo\.com\/.+/
export const MATCH_VIMEO_LIVESTREAM = /vimeo\.com\/event\/.+/
export const MATCH_TED = /embed\.ted\.com\/.+/
export const MATCH_FACEBOOK = /^https?:\/\/(www\.)?facebook\.com.*\/(video(s)?|watch|story)(\.php?|\/).+$/
export const MATCH_FACEBOOK_WATCH = /^https?:\/\/fb\.watch\/.+$/
export const MATCH_STREAMABLE = /streamable\.com\/([a-z0-9]+)$/
export const MATCH_MIX_CLOUD = /mixcloud\.com\/([^/]+\/[^/]+)/

export const AUDIO_EXTENSIONS = /\.(m4a|mp4a|mpga|mp2|mp2a|mp3|m2a|m3a|wav|weba|aac|oga|spx)($|\?)/i
export const VIDEO_EXTENSIONS = /\.(mp4|og[gv]|webm|mov|m4v)($|\?)/i
export const HLS_EXTENSIONS = /\.(m3u8)($|\?)/i
export const DASH_EXTENSIONS = /\.(mpd)($|\?)/i
export const FLV_EXTENSIONS = /\.(flv)($|\?)/

const canPlayFile = url => {
  if (url instanceof Array) {
    for (const item of url) {
      if (typeof item === 'string' && canPlayFile(item)) {
        return true
      }
      if (canPlayFile(item.src)) {
        return true
      }
    }
    return false
  }

  return (
    AUDIO_EXTENSIONS.test(url) ||
    VIDEO_EXTENSIONS.test(url) ||
    HLS_EXTENSIONS.test(url) ||
    DASH_EXTENSIONS.test(url) ||
    FLV_EXTENSIONS.test(url)
  )
}

export const canPlay = {
  youtube: url => {
    if (url instanceof Array) {
      return url.every(item => MATCH_YOUTUBE.test(item))
    }
    return MATCH_YOUTUBE.test(url)
  },
  soundcloud: url => MATCH_SOUNDCLOUD.test(url) && !AUDIO_EXTENSIONS.test(url),
  vimeo: url =>
    MATCH_VIMEO.test(url) &&
    !MATCH_VIMEO_LIVESTREAM.test(url) &&
    !VIDEO_EXTENSIONS.test(url) &&
    !HLS_EXTENSIONS.test(url),
  livestream: url =>
    MATCH_VIMEO_LIVESTREAM.test(url) && !VIDEO_EXTENSIONS.test(url) && !HLS_EXTENSIONS.test(url),
  ted: url => MATCH_TED.test(url) && !VIDEO_EXTENSIONS.test(url) && !HLS_EXTENSIONS.test(url),
  facebook: url => MATCH_FACEBOOK.test(url) || MATCH_FACEBOOK_WATCH.test(url),
  streamable: url => MATCH_STREAMABLE.test(url),
  mixcloud: url => MATCH_MIX_CLOUD.test(url),
  file: canPlayFile,
}

export const format = seconds => {
  const date = new Date(seconds * 1000)
  const hh = date.getUTCHours()
  const mm = date.getUTCMinutes()
  const ss = pad(date.getUTCSeconds())
  if (hh) {
    return `${hh}:${pad(mm)}:${ss}`
  }
  return `${mm}:${ss}`
}

const pad = string => {
  return ('0' + string).slice(-2)
}

export const duration = duration => format(duration)

export const elapsed = (duration, played) => format(duration * played)

export const remaining = (duration, played) => format(duration * (1 - played))
