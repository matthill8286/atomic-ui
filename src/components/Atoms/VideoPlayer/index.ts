export * from './VideoPlayerCore'
export * from './VideoPlayerWrapper'
export * from './VideoPlayer.interface'
export * from './helpers'
