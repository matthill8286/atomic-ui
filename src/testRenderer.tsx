import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
