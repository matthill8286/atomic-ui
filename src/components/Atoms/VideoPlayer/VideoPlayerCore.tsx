import React, {
  MutableRefObject,
  ReactElement,
  Ref,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react'
import { useInView } from 'react-intersection-observer'
import { SkeletonBlockItem, SkeletonInlineItem } from '@/components/Atoms/Skeleton'
import {
  VideoPlayerProps,
  VideoPlayerViewer,
} from '@/components/Atoms/VideoPlayer/VideoPlayer.interface'
import {
  StyledVideoPlayer,
  StyledVideoWrapper,
} from '@/components/Atoms/VideoPlayer/VideoPlayer.styled'

export const VideoPlayerCore: React.FC<VideoPlayerProps> = ({
  title = 'core video player',
  videoId = 'player-core',
  width = '100%',
  height = '100%',
  url,
  pip = false,
  controls = true,
  light = false,
  loop = false,
  loading,
  playbackRate = 1,
  playing,
  volume = 0.6,
  muted = false,
  allowfullscreen = false,
  rounded = false,
  stopOnUnmount = true,
  maxHeight,
  onStart,
  onPause,
  onPlay,
  onError,
  onEnded,
  onProgress,
  onDuration,
  children,
  ...props
}) => {
  if (loading) {
    return <SkeletonBlockItem height="350px" />
  }

  const containerRef = useRef<HTMLDivElement>(null)
  const [videoViewer, setVideoViewer] = useState<VideoPlayerViewer | null>(null)
  const [ref, inView] = useInView({
    threshold: 0.3,
  })

  const [loaded, setLoaded] = useState(false)

  const id = `video-player-container-${videoId}`

  const _onStart = useCallback(cb => onStart?.(cb), [])

  const _onEnded = useCallback(event => onEnded?.(event), [])

  const _onPlay = useCallback(event => onPlay?.(event), [])

  const _onDuration = useCallback(event => onDuration?.(event), [])

  const _onPause = useCallback(event => onPause?.(event), [])

  const _onProgress = useCallback(state => {
    onProgress?.(state)
  }, [])

  const _onError = useCallback(event => onError?.(event), [])

  useEffect(() => {
    videoViewer?.onStart(() => {
      if (!loaded) {
        setLoaded(true)
      }
    })
  }, [videoViewer])

  useEffect(() => {
    if (
      document.fullscreenElement !== null &&
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      (document.fullscreenElement === true ||
        (document.fullscreenElement as HTMLVideoElement)?.canPlayType !== undefined)
    ) {
      return
    }
    if (loaded && !inView) {
      videoViewer?.stop()
    }
  }, [inView])

  return (
    <StyledVideoWrapper aspect="16/9" ref={containerRef}>
      <StyledVideoPlayer
        id={id}
        width={width}
        ref={videoViewer}
        height={height}
        url={url}
        pip={pip}
        playing={playing}
        controls={controls}
        rounded={rounded ?? undefined}
        light={light}
        loop={loop}
        playbackRate={playbackRate}
        volume={volume}
        muted={muted}
        onStart={_onStart}
        onPlay={_onPlay}
        onPause={_onPause}
        onProgress={_onProgress}
        onError={_onError}
        onDuration={_onDuration}
      />
      <div id={id} ref={ref} />
    </StyledVideoWrapper>
  )
}
