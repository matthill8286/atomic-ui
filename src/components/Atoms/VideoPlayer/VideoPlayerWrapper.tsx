import React from 'react'
import { SkeletonBlockItem } from '@/components/Atoms/Skeleton'
import { VideoPlayerProps } from '@/components/Atoms/VideoPlayer/VideoPlayer.interface'
import { EmbeddedWrapper } from '@/components/Atoms/EmbeddedPlayer'
import { VideoPlayerCore } from '@/components/Atoms/VideoPlayer/VideoPlayerCore'

export const VideoPlayerWrapper: React.FC<VideoPlayerProps> = ({
  title = 'core video player',
  unsupportedMedia,
  context = 'page',
  videoId = 'player-core',
  width = '100%',
  height = '100%',
  url,
  pip = false,
  controls = true,
  light = false,
  loop = false,
  loading,
  playbackRate = 1,
  playing,
  volume = 0.6,
  muted = false,
  allowfullscreen = false,
  rounded = false,
  stopOnUnmount = true,
  maxHeight,
  onStart,
  onPause,
  onPlay,
  onError,
  onEnded,
  onProgress,
  onDuration,
  children,
  ...props
}) => {
  if (loading) {
    return <SkeletonBlockItem height="350px" />
  }

  if (unsupportedMedia) {
    return (
      <EmbeddedWrapper
        height={height}
        aspect="16/9"
        width={width}
        embedUrl={url}
        title={title}
        allowfullscreen={false}
      />
    )
  }

  return (
    <VideoPlayerCore
      videoId={videoId}
      width={width}
      height={height}
      url={url}
      pip={pip}
      playing={playing}
      controls={controls}
      rounded={rounded ?? undefined}
      light={light}
      loop={loop}
      playbackRate={playbackRate}
      volume={volume}
      muted={muted}
      onStart={onStart}
      onPlay={onPlay}
      onPause={onPause}
      onProgress={onProgress}
      onError={onError}
      onDuration={onDuration}
    />
  )
}
