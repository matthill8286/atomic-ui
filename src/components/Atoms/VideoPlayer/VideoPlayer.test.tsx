import React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { VideoPlayerCore } from '@/components/Atoms/VideoPlayer/VideoPlayerCore'

describe('VideoPlayer', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(
      <VideoPlayerCore url="https://www.youtube.com/embed/-BdbiZcNBXg" />
    )
    expect(tree).toMatchSnapshot()
  })
})
