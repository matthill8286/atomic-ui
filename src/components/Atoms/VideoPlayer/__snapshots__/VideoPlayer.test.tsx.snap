// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer renders correctly 1`] = `
.c0 {
  position: relative;
  height: 100%;
  width: 100%;
}

.c0:before {
  content: '';
  display: block;
  padding-bottom: calc(100% / calc(16/9));
}

.c0 > img {
  height: auto;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  border: none;
}

<div
  className="c0"
>
  <div
    className="c1"
    id="video-player-container-player-core"
    rounded={false}
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
  <div
    id="video-player-container-player-core"
  />
</div>
`;
